### PR TITLE
executor: grant to non-existent table with at least create privilege

### LINKED ISSUE #29268: Grant SELECT and INSERT to non-existent table should report an error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

This is really confusing in MySQL:
When granting `ALTER`, or `INSERT, CREATE` privileges, it works, see https://github.com/pingcap/tidb/issues/28533
However, when granting `INSERT` or `SELECT`, it doesn't work.

```sql
mysql> drop table if exists tmpdb.t1;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> grant select on tmpdb.t1 to test_user;
Query OK, 0 rows affected (0.18 sec)

mysql> grant insert on tmpdb.t1 to test_user;
Query OK, 0 rows affected (0.18 sec)
```

### 2. What did you expect to see? (Required)

```sql
mysql> drop table if exists tmpdb.t1;
Query OK, 0 rows affected, 1 warning (0.01 sec)

mysql> grant select on tmpdb.t1 to test_user;
ERROR 1146 (42S02): Table 'tmpdb.t1' doesn't exist
mysql> grant insert on tmpdb.t1 to test_user;
ERROR 1146 (42S02): Table 'tmpdb.t1' doesn't exist
```

### 3. What did you see instead (Required)

```sql
mysql> drop table if exists tmpdb.t1;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> grant select on tmpdb.t1 to test_user;
Query OK, 0 rows affected (0.18 sec)

mysql> grant insert on tmpdb.t1 to test_user;
Query OK, 0 rows affected (0.18 sec)
```

### 4. What is your TiDB version? (Required)

```
Release Version: None
Edition: Community
Git Commit Hash: None
Git Branch: None
UTC Build Time: None
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by unconsolable**
Signed-off-by: unconsolable <chenzhipeng2012@gmail.com>

### What problem does this PR solve?

Issue Number: close #29268 

Problem Summary:

### What is changed and how it works?
- When grant to non-existent table, check if the one of the privileges is create or all.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support grant to non-existent table with at least create privilege to keep compatibility with MySQL 8.0
```

### CODE REVIEW

File: executor/grant.go [hunk 1]
```diff
@@ -77,9 +77,20 @@ func (e *GrantExec) Next(ctx context.Context, req *chunk.Chunk) error {
 		dbNameStr := model.NewCIStr(dbName)
 		schema := e.ctx.GetInfoSchema().(infoschema.InfoSchema)
 		tbl, err := schema.TableByName(dbNameStr, model.NewCIStr(e.Level.TableName))
-		// Allow GRANT on non-existent table, see issue #28533
-		if err != nil && !terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
-			return err
+		// Allow GRANT on non-existent table with at least create privilege, see issue #28533 #29268
+		if err != nil {
```
Comment (lcwangchao) [inline executor/grant.go:81]: Will this pr cause some compatibility problems?

Comment (unconsolable) [reply executor/grant.go:81]: IMHO it won't as it is based on previous PR https://github.com/pingcap/tidb/pull/28882, and both of them have not delivered to release branch. Or we can wait on @morgo's opinion.

File: executor/grant_test.go [hunk 2]
```diff
@@ -474,6 +474,38 @@ func TestGrantOnNonExistTable(t *testing.T) {
...
+	defer tk.MustExec("DROP DATABASE IF EXISTS d29268")
+	tk.MustExec("USE d29268")
+	tk.MustExec("CREATE USER u29268")
+	defer tk.MustExec("DROP USER u29268")
+
+	// without create privilege
+	err = tk.ExecToErr("GRANT SELECT ON t29268 TO u29268")
+	require.Error(t, err)
+	require.True(t, terror.ErrorEqual(err, infoschema.ErrTableNotExists))
+	fmt.Println(err)
```
Comment (AilinKid) [inline executor/grant_test.go:489]: better remove this line

Comment (unconsolable) [reply executor/grant_test.go:489]: Forgot to remove when developing, I will remove it and check others.

File: executor/grant_test.go [hunk 3]
```diff
@@ -474,6 +474,37 @@ func TestGrantOnNonExistTable(t *testing.T) {
...
+	require.True(t, terror.ErrorEqual(err, infoschema.ErrTableNotExists))
+	err = tk.ExecToErr("GRANT INSERT, REFERENCES, ALTER ON t29268 TO u29268")
+	require.Error(t, err)
+	require.True(t, terror.ErrorEqual(err, infoschema.ErrTableNotExists))
+
+	// with create privilege
+	tk.MustExec("GRANT CREATE ON t29268 TO u29268")
+	tk.MustExec("GRANT CREATE, SELECT ON t29268 TO u29268")
+	tk.MustExec("GRANT CREATE, DROP, CREATE VIEW ON t29268 TO u29268")
+	tk.MustExec("GRANT CREATE, INSERT, REFERENCES, ALTER ON t29268 TO u29268")
```
Comment (djshow832) [inline executor/grant_test.go:506]: Are all the privileges take effect after the table is created? Please try it in MySQL and add tests here.

Comment (unconsolable) [reply executor/grant_test.go]: Yes, it works after creation. Test cases added.

**Review by djshow832 — APPROVED** [review]

File: executor/grant.go (hunk 1 above)
Comment (morgo) [reply executor/grant.go:81]: I agree with you @unconsolable. It will only cause compatibility issues between from an unreleased version.

**Review by morgo — APPROVED** [review]

**Pull request merged**